client/ipa-join.c

```diff
@@ -612,161 +632,417 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
 
-    if (server) {
-        ipaserver = strdup(server);
-    } else {
-        char * conf_data = read_config_file(IPA_CONFIG);
-        if ((ipaserver = getIPAserver(conf_data)) == NULL) {
-            if (!quiet)
-                fprintf(stderr, _("Unable to determine IPA server from %s\n"),
-                        IPA_CONFIG);
-            exit(1);
-        }
-        free(conf_data);
+        rval = 17;
+        goto cleanup;
     }
 
-    if (NULL == hostname) {
-        uname(&uinfo);
-        host = strdup(uinfo.nodename);
-    } else {
-        host = strdup(hostname);
+    curl = curl_easy_init();
+    if (!curl) {
+        if (!quiet)
+            fprintf(stderr, _("curl_easy_init() failed\n"));
+
+        rval = 17;
+        goto cleanup;
     }
 
-    if (NULL == host) {
-        rval = 3;
+    /* setting endpoint and custom headers */
+    ASPRINTF(&url, "https://%s/ipa/json", ipaserver);
+    CURL_SETOPT(curl, CURLOPT_URL, url);
+
+    ASPRINTF(&referer, "referer: https://%s/ipa", ipaserver);
+    headers = curl_slist_append_log(headers, referer, quiet);
+    if (!headers) {
+        rval = 17;
         goto cleanup;
     }
 
-    if (NULL == strstr(host, ".")) {
-        if (!quiet)
-            fprintf(stderr, _("The hostname must be fully-qualified: %s\n"),
-                    host);
-        rval = 16;
+    ASPRINTF(&user_agent, "User-Agent: %s/%s", NAME, VERSION);
+    headers = curl_slist_append_log(headers, user_agent, quiet);
+    if (!headers) {
+        rval = 17;
         goto cleanup;
     }
 
-    krberr = krb5_init_context(&krbctx);
-    if (krberr) {
-        if (!quiet)
-            fprintf(stderr, _("Unable to join host: "
-                              "Kerberos context initialization failed\n"));
-        rval = 1;
+    headers = curl_slist_append_log(headers, "Accept: application/json", quiet);
+    if (!headers) {
+        rval = 17;
         goto cleanup;
     }
-    krberr = krb5_kt_resolve(krbctx, ktname, &keytab);
-    if (krberr != 0) {
-        if (!quiet)
-            fprintf(stderr, _("Error resolving keytab: %s.\n"),
-                error_message(krberr));
-        rval = 7;
+
+    headers = curl_slist_append_log(headers, "Content-Type: application/json", quiet);
+    if (!headers) {
+        rval = 17;
         goto cleanup;
     }
+    CURL_SETOPT(curl, CURLOPT_HTTPHEADER, headers);
 
-    krberr = krb5_get_default_realm(krbctx, &realm);
-    if (krberr != 0) {
-        if (!quiet)
-            fprintf(stderr, _("Error getting default Kerberos realm: %s.\n"),
-                error_message(krberr));
-        rval = 21;
+    CURL_SETOPT(curl, CURLOPT_CAINFO, DEFAULT_CA_CERT_FILE);
+
+    CURL_SETOPT(curl, CURLOPT_WRITEFUNCTION, &jsonrpc_handle_response);
+    CURL_SETOPT(curl, CURLOPT_WRITEDATA, response);
+
+    /* delegate authentication to GSSAPI */
+    CURL_SETOPT(curl, CURLOPT_GSSAPI_DELEGATION, CURLGSSAPI_DELEGATION_FLAG);
+    CURL_SETOPT(curl, CURLOPT_HTTPAUTH, CURLAUTH_NEGOTIATE);
```

Comment:
I have the feeling this is delegating the TGT. xmlrpc-c enables it with (simplified):

```
    rc = curl_easy_setopt(curlSessionP, CURLOPT_HTTPAUTH, CURLAUTH_GSSNEGOTIATE)
```

Reply:
`CURLAUTH_GSSNEGOTIATE` has been deprecated in https://github.com/curl/curl/commit/e38ba430142153e95e2b0abe08a31d626a5c5864 and is an alias for `CURLAUTH_NEGOTIATE`.

Reply:
Ok I have some additional changes but I'll submit them as a separate PR.